### PR TITLE
Support @ConfigMapping in Extensions

### LINKED ISSUE #7856: @ConfigItem Optional<Map<String,String>> does not work
Declaring a config as: 

    /**
     * Sets the placeholders to replace in SQL migration scripts.
     */
    @ConfigItem
    public Optional<Map<String, String>> placeholders = Optional.empty();

Always returns an `Optional.empty()` when having the following in the `application.properties`: 

```
quarkus.flyway.placeholders.foo=bar
```

### LINKED ISSUE #7862: Optional configuration groups do not work
**Describe the bug**
When I was working on updating `quarkus-oidc` OidcTenantResolver to support different client secret methods, we ended agreeing with @pedroigor on the following as an initial step:
```
@ConfigGroup
class Credentials {
    @ConfigItem
    Optional<String> secret;
    
    @ConfigItem
    Secret clientSecret = new Secret();

    @ConfigGroup
    public class Secret {
        @ConfigItem
        Optional<String> value;
        // etc
    } 
    // there will more groups like `Secret` here for other authentication methods
}
``` 
Note `clientSecret` has to be initialized immediately and `Secret` properties are optional, because all immediate `Credentials` properties (`secret`, `clientSecret` and more later on) are mutually exclusive.

I tried:
```
class Credentials {
    @ConfigItem
    Optional<String> secret;
    
    @ConfigItem
    Optional<Secret> clientSecret;

    @ConfigGroup
    public class Secret {
        @ConfigItem
        // non-optional
        String value;
        // etc
    } 
}
```

which looks nicer IMHO, and would've let me to just do `if clientSecret.isPresent()` instead of checking all of `Secret` properties if any of them has been unnecessarily set.

**Expected behavior**
It would super great if optional configuration groups were supported
**Actual behavior**
When I tried `Optional<Secret> clientSecret` and configured
```
quarkus.oidc.credentials.client-secret.value=mysecret
```
I got an exception (it was a class cast to `Optional`), I can find out....
It can be tricky to fix I guess, may be if at least one of a given optional group's properties is set then initialize and treat all other non initialized required properties in this group as optional because the group itself is optional 

**To Reproduce**
See above

### LINKED ISSUE #9791: Add "prefix" to @ConfigRoot
**Description**
We are using Quarkus at a large scale to develop many microservices. In this setup we also have quite a few homegrown libraries and tools, which we now look to integrate into the Quarkus development as Quarkus extensions. For this purpose we want to use `@ConfigRoot` et al. to define the configuration properties and also generate corresponding documentation, as is being done for Quarkus itself. Since most of the extensions we are developing are not applicable outside our corporation, these won't be opensourced under the Quarkus project and so it doesn't really make any sense for the configuration properties to have the `quarkus` prefix.

I would therefor like to propose that the `@ConfigRoot` annotation type is extended with a `prefix` attribute (possibly defaulting to `"quarkus"`, so that we can specify our own prefix.

**Implementation ideas**
As described, I think a new `prefix` attribute on `@ConfigRoot` would make sense.

### LINKED ISSUE #17088: The optional configuration group is always empty.
## Describe the bug
I customized a quarkus extension of the vertx web client and configured WebClient instantiation parameters through the application.properties file. The proxy configuration is always empty, it is an optional configuration group.

RootWebClientConfiguration.java
```java
@ConfigRoot(name = "spinach.web.client", phase = ConfigPhase.RUN_TIME)
public class RootWebClientConfiguration {

    public static final String DEFAULT_CLIENT = "<default>";

    public static boolean isDefault(String name) {
        return DEFAULT_CLIENT.equals(name);
    }

    /**
     * defaultConfiguration
     */
    @ConfigItem(name = ConfigItem.PARENT)
    public WebClientConfiguration defaultConfiguration;

    /**
     * additionalConfiguration
     */
    @ConfigItem(name = ConfigItem.PARENT)
    public Map<String, WebClientConfiguration> additionalConfiguration;

    public WebClientConfiguration getWebClientConfiguration(String name) {
        return isDefault(name) ? defaultConfiguration : additionalConfiguration.get(name);
    }
}
```

WebClientConfiguration.java
```java
@ConfigGroup
public class WebClientConfiguration {

    /**
     * userAgentEnabled
     */
    @ConfigItem(defaultValue = "false")
    boolean userAgentEnabled;

    /**
     * userAgent
     */
    @ConfigItem
    Optional<String> userAgent;

    /**
     * maxPoolSize
     */
    @ConfigItem(defaultValue = "16")
    int maxPoolSize;

    /**
     * tryUseCompression
     */
    @ConfigItem(defaultValue = "true")
    boolean tryUseCompression;

    /**
     * keepAlive
     */
    @ConfigItem(defaultValue = "true")
    boolean keepAlive;

    /**
     * keepAliveTimeout
     */
    @ConfigItem(defaultValue = "60s")
    Duration keepAliveTimeout;

    /**
     * connectTimeout
     */
    @ConfigItem(defaultValue = "60s")
    Duration connectTimeout;

    /**
     * protocolVersion
     */
    @ConfigItem
    @ConvertWith(HttpVersionConverter.class)
    Optional<HttpVersion> protocolVersion;

    /**
     * maxWaitQueueSize
     */
    @ConfigItem(defaultValue = "-1")
    int maxWaitQueueSize;

    /**
     * followRedirects
     */
    @ConfigItem(defaultValue = "true")
    boolean followRedirects;

    /**
     * tcpNoDelay
     */
    @ConfigItem(defaultValue = "true")
    boolean tcpNoDelay;

    /**
     * tcpFastOpen
     */
    @ConfigItem(defaultValue = "false")
    boolean tcpFastOpen;

    /**
     * tcpKeepAlive
     */
    @ConfigItem(defaultValue = "true")
    boolean tcpKeepAlive;

    /**
     * tcpCork
     */
    @ConfigItem(defaultValue = "false")
    boolean tcpCork;

    /**
     * tcpQuickAck
     */
    @ConfigItem(defaultValue = "false")
    boolean tcpQuickAck;

    /**
     * http2MaxPoolSize
     */
    @ConfigItem(defaultValue = "8")
    int http2MaxPoolSize;

    /**
     * http2KeepAliveTimeout
     */
    @ConfigItem(defaultValue = "60s")
    Duration http2KeepAliveTimeout;

    /**
     * trustAll
     */
    @ConfigItem
    boolean trustAll;

    /**
     * verifyHost
     */
    @ConfigItem
    boolean verifyHost;

    /**
     * proxyConfiguration
     */
    @ConfigItem
    Optional<ProxyConfiguration> proxy;

    public static class HttpVersionConverter implements Converter<HttpVersion> {

        @Override
        public HttpVersion convert(String value) {
            value = value.trim().toUpperCase();
            switch (value) {
                case "HTTP_1_1":
                    return HttpVersion.HTTP_1_1;
                case "HTTP_1_0":
                    return HttpVersion.HTTP_1_0;
                case "HTTP_2":
                    return HttpVersion.HTTP_2;
                default:
                    throw new IllegalArgumentException("Unsupported value [" + value + "] given.");
            }
        }
    }
}
```

ProxyConfiguration.java
```java
@ConfigGroup
public class ProxyConfiguration {

    /**
     * type
     */
//    @DefaultConverter
    @ConvertWith(ProxyTypeConverter.class)
    @ConfigItem
    ProxyType type;

    /**
     * host
     */
    @ConfigItem
    String host;

    /**
     * port
     */
    @ConfigItem
    int port;

    /**
     * username
     */
    @ConfigItem
    Optional<String> username;

    /**
     * password
     */
    @ConfigItem
    Optional<String> password;

    public static class ProxyTypeConverter implements Converter<ProxyType> {

        @Override
        public ProxyType convert(String value) {
            if (value == null || value.isEmpty()) {
                throw new IllegalArgumentException("Proxy type is empty.");
            }
            value = value.trim().toUpperCase();
            switch (value) {
                case "HTTP":
                    return ProxyType.HTTP;
                case "SOCKS5":
                    return ProxyType.SOCKS5;
                case "SOCKS4":
                    return ProxyType.SOCKS4;
                default:
                    throw new IllegalArgumentException("Unsupported value [" + value + "] given.");
            }
        }
    }

}
```

application.properties

```properties 
quarkus.spinach.web.client.proxy.type=HTTP
quarkus.spinach.web.client.proxy.host=127.0.0.1
quarkus.spinach.web.client.proxy.port=1080
quarkus.spinach.web.client.max-pool-size=32
```

### Expected behavior

Proxy parameter is not empty.

### Actual behavior

Proxy parameter is empty.

## To Reproduce

[https://github.com/aomsweet/spinach-quarkus-extension](https://github.com/aomsweet/spinach-quarkus-extension)

1.  mvn clean install
2.  cd app-examples
3.  mvn quarkus:dev
4.  open http://localhost:8080/config

Then you can see: `"proxy":{"value":null,"present":false}`

### Configuration
```properties
# Add your application.properties here, if applicable.

quarkus.spinach.web.client.proxy.type=HTTP
quarkus.spinach.web.client.proxy.host=127.0.0.1
quarkus.spinach.web.client.proxy.port=1080
quarkus.spinach.web.client.max-pool-size=32

```

### Screenshots
image

### Output of `java -version`

JDK: 1.8

### GraalVM version (if different from Java)

GraalVM: 21.1.0

### Quarkus version or git rev

Quarkus: 1.13.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

maven: 3.8.1

## Additional context
(Add any other context about the problem here.)

### PR DESCRIPTION
**Pull request by radcortez**
- [x] Requires https://github.com/smallrye/smallrye-config/pull/626
- [x] Requires https://github.com/quarkusio/quarkus/pull/20168
- [x] Requires https://github.com/quarkusio/quarkus/pull/20103

Fixes the following `@ConfigRoot` issues by allowing to use `@ConfigMapping` instead:
- Fixes https://github.com/quarkusio/quarkus/issues/7856
- Fixes https://github.com/quarkusio/quarkus/issues/7862
- Fixes https://github.com/quarkusio/quarkus/issues/9791
- Fixes https://github.com/quarkusio/quarkus/issues/17088

Does not support constructor injection in Recorders. This can be added later.

### CODE REVIEW

**Comment by gsmet** [comment]
@radcortez this one fell through the cracks. Is it still needed? If so it needs a big rebase.

**Comment by radcortez** [comment]
> @radcortez this one fell through the cracks. Is it still needed? If so it needs a big rebase.

Yes, it will give us the ability to use SR `@ConfigMapping` for extensions. It allows us to create config structures not currently possible with the `@ConfigRoot` and in a VERY distant future consolidate on a single configuration mechanism :)

File: core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java [hunk 1]
```diff
@@ -49,7 +50,7 @@ public ConfigDocItemScanner() {
      */
     public void addConfigGroups(TypeElement configGroup) {
         String configGroupName = configGroup.getQualifiedName().toString();
-        if (configGroupName.startsWith(IO_QUARKUS_TEST_EXTENSION_PACKAGE)) {
+        if (!configGroupName.startsWith(IO_QUARKUS_TEST_EXTENSION_PACKAGE)) {
```
Comment (gastaldi) [inline core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java:53]: Is this change and the one below intentional?

Comment (radcortez) [reply core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDocItemScanner.java]: Very good catch. I knew I was going to forget about this! (it was to test stuff).

**Review by gastaldi — APPROVED** [review]

**Comment by radcortez** [comment]
Uh oh!

**Pull request merged**